Python/Product/PythonTools/PythonTools/Environments/AddVirtualEnvironmentView.cs

```diff
@@ -0,0 +1,521 @@
+﻿// Python Tools for Visual Studio
+// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Windows;
+using Microsoft.PythonTools.Infrastructure;
+using Microsoft.PythonTools.Interpreter;
+using Microsoft.PythonTools.Project;
+
+namespace Microsoft.PythonTools.Environments {
+    sealed class AddVirtualEnvironmentView : EnvironmentViewBase, INotifyPropertyChanged {
```

Comment:
It isn't a good idea to have mix of `DependencyObject` and `INotifyPropertyChanged`. I'd actually prefer to have the latter: it is easier to debug, more lightweight, better handled inside collections and not that slow in recent versions of WPF

Reply:
Turns out it's not used anymore, it was part of the old virtual env view class. I'll remove it as a base.

As far as switching everything to `INotifyPropertyChanged`, this is too risky / last minute for this upcoming release. I agree with you though, there isn't much benefit to DependencyObject, and it's a lot more line sof code.

Reply:
The SafeGetValue/SafeSetValue also appear to not be needed. All property access is done from main thread